components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,665 @@
+//!     (4) signal                               |   |                 |
+//! +--------------+  (3) PrepareMergeResult  +--+---v------+          |
+//! | source apply +--------------------------> source peer <----------+
+//! +--------------+                          +-------------+
+//! ```
+//!
+//! At first, target region will not apply the `CommitMerge` command. Instead
+//! the apply progress will be paused and it redirects the log entries from
+//! source region, as a `CatchUpLogs` message, to the local source region peer
+//! (step 2). When the source region peer has applied all logs up to the prior
+//! `PrepareMerge` command, it will send its tablet to the target region (step
+//! 4). Here we use a temporary channel instead of directly sending message
+//! between apply FSMs like in v1.
+//!
+//! ## On Apply Result (`Peer::on_apply_res_commit_merge`)
+//!
+//! Update the target peer states and send a `MergeResult` message to source
+//! peer to destroy it.
+
+use std::{any::Any, cmp, path::PathBuf};
+
+use crossbeam::channel::SendError;
+use engine_traits::{KvEngine, RaftEngine, RaftLogBatch, TabletContext};
+use futures::channel::oneshot;
+use kvproto::{
+    metapb::{self, Region},
+    raft_cmdpb::{AdminCmdType, AdminRequest, AdminResponse, CommitMergeRequest, RaftCmdRequest},
+    raft_serverpb::PeerState,
+};
+use protobuf::Message;
+use raft::{GetEntriesContext, Storage, INVALID_ID, NO_LIMIT};
+use raftstore::{
+    coprocessor::RegionChangeReason,
+    store::{
+        fsm::new_admin_request, metrics::PEER_ADMIN_CMD_COUNTER, util, ProposalContext, Transport,
+    },
+    Result,
+};
+use slog::{debug, error, info, warn};
+use tikv_util::{
+    box_err, slog_panic,
+    store::{find_peer, region_on_same_stores},
+};
+
+use super::merge_source_path;
+use crate::{
+    batch::StoreContext,
+    fsm::ApplyResReporter,
+    operation::{AdminCmdResult, SharedReadTablet},
+    raft::{Apply, Peer},
+    router::{CmdResChannel, PeerMsg, PeerTick, StoreMsg},
+};
+
+#[derive(Debug)]
+pub struct CommitMergeResult {
+    pub index: u64,
+    source_path: PathBuf,
+    region: Region,
+    source: Region,
+    tablet: Box<dyn Any + Send + Sync>,
+}
+
+#[derive(Debug)]
+pub struct CatchUpLogs {
+    pub target_region_id: u64,
+    pub merge: CommitMergeRequest,
+    pub tx: oneshot::Sender<()>,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    // Match v1::on_check_merge. Called on source peer.
+    pub fn on_check_merge<T: Transport>(&mut self, store_ctx: &mut StoreContext<EK, ER, T>) {
+        if !self.serving() || self.applied_merge_state().is_none() {
+            return;
+        }
+        self.add_pending_tick(PeerTick::CheckMerge);
+        if let Err(e) = self.schedule_commit_merge_proposal(store_ctx) {
+            info!(
+                self.logger,
+                "failed to schedule commit merge";
+                "err" => ?e,
+            );
+            // TODO: self.handle_commit_merge_failure(store_ctx, e);
```

Comment:
Now rollback should only be triggered on rejection.

Reply:
If expected target is stale, this should be a fast path to rollback instead of fetching entries.

Reply:
Rollback is unlikely, only processing on one path is simpler and more clear.

Reply:
There're other errors, like the error returned by `calculate_min_progress`. If you just want to reuse the code, we can always use the same function later.